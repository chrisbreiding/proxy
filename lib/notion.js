const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
