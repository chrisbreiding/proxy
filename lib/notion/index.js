const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
