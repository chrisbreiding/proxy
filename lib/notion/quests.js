const {
  getBlockChildren,
  getPlainText,
} = require('./util')

const findUpcomingId = async (blocks) => {
  const block = blocks.find((block) => {
    return block.type === 'toggle' && getPlainText(block) === 'Upcoming'
  })

  if (!block) {
    throw new Error('Could not find Upcoming block')
  }

  return block.id
}

const getAll = async ({ notionToken, pageId }) => {
  const { results: questBlocks } = await getBlockChildren({ notionToken, pageId })
  const upcomingId = await findUpcomingId(questBlocks)
  const { results: upcomingBlocks } = await getBlockChildren({ notionToken, pageId: upcomingId })

  return [...questBlocks, ...upcomingBlocks]
}

module.exports = {
  getAll,
}
