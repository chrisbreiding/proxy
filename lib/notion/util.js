const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
