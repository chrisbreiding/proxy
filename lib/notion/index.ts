import type express from 'express'
import { addMeal as addFactorMeal, getMeals as getFactorMeals } from './factor/meals'
import { addQuest } from './quests'
import { onSocket } from './shopping'
import { addUpcomingWeek } from './upcoming-week'
import { getBlockChildren } from './util/queries'
import { clearCompleted, deleteRecentlyCleared } from './clear-completed'
import { sendHtml } from './util/general'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export function action (req: express.Request, res: express.Response) {
  const action = req.query.action

  if (!action) {
    sendHtml(res, 400, '<p>A value for <em>action</em> must be provided in the query string</p>')

    return
  }

  switch (action) {
    case 'addUpcomingWeek':
      return addUpcomingWeek(req, res)
    case 'clearCompleted':
      return clearCompleted(req, res)
    case 'deleteRecentlyCleared':
      return deleteRecentlyCleared(req, res)
    default:
      sendHtml(res, 400, `<p>Action not supported: ${action}</p>`)
  }
}

export {
  addFactorMeal,
  addQuest,
  getFactorMeals,
  onSocket,
}
