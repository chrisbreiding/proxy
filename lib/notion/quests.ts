import { getBlockChildren, getBlockPlainText, NotionBlock } from './util'

async function findUpcomingId (blocks: NotionBlock[]) {
  const block = blocks.find((block) => {
    return block.type === 'toggle' && getBlockPlainText(block) === 'Upcoming'
  })

  if (!block) {
    throw new Error('Could not find Upcoming block')
  }

  return block.id
}

interface GetAllOptions {
  notionToken: string
  pageId: string
}

export async function getAllQuests ({ notionToken, pageId }: GetAllOptions) {
  const questBlocks = await getBlockChildren({ notionToken, pageId })
  const upcomingId = await findUpcomingId(questBlocks)
  const upcomingBlocks = await getBlockChildren({ notionToken, pageId: upcomingId })

  return [...questBlocks, ...upcomingBlocks]
}
