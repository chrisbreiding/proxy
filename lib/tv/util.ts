import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
