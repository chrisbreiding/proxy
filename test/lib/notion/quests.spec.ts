import fs from 'fs-extra'
import nock from 'nock'
import { afterEach, describe, expect, it } from 'vitest'

import { fixture, nockGetBlockChildren } from '../../support/util'
import { getAll } from '../../../lib/notion/quests'

describe('lib/notion/quests', () => {
  afterEach(() => {
    nock.cleanAll()
  })

  describe('#getAll', () => {
    it('returns all quest blocks', async () => {
      nockGetBlockChildren('page-id', { fixture: 'quests/quests-blocks' })
      nockGetBlockChildren('upcoming-id', { fixture: 'quests/upcoming-blocks' })

      const result = await getAll({
        notionToken: 'notion-token',
        pageId: 'page-id',
      })

      const expected = fs.readJsonSync(fixture('quests/get-all-result'))

      expect(result).to.deep.equal(expected)
    })

    it('errors if no upcoming block found', async () => {
      const questBlocks = fs.readJsonSync(fixture('quests/quests-blocks'))

      questBlocks.results = questBlocks.results.slice(0, questBlocks.results.length - 1)

      nockGetBlockChildren('page-id', { reply: questBlocks })

      try {
        await getAll({
          notionToken: 'notion-token',
          pageId: 'page-id',
        })

        throw new Error('Expected upcoming not to be found')
      } catch (error: any) {
        expect(error.message).to.equal('Could not find Upcoming block')
      }
    })
  })
})
