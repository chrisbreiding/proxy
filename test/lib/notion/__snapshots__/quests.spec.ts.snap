// Vitest Snapshot v1

exports[`lib/notion/quests > #getAll > returns all quest blocks 1`] = `
[
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 1 task one",
          "text": {
            "content": "Day 1 task one",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "aeb55ad0-de19-4f6a-8c8a-b00b6b006e25",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 1 task two",
          "text": {
            "content": "Day 1 task two",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "7ff8f624-1aa1-4cec-9a2f-b0d4d0cffed5",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Mon, 10/31    ‚òîÔ∏è ",
          "text": {
            "content": "Mon, 10/31    ‚òîÔ∏è ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "(55%) ",
          "text": {
            "content": "(55%) ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "blue",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "57¬∞",
          "text": {
            "content": "57¬∞",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": " / ",
          "text": {
            "content": " / ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "orange",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "65¬∞",
          "text": {
            "content": "65¬∞",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "d6679deb-4419-4d39-8e63-5a65a5c3d3de",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 2 task one",
          "text": {
            "content": "Day 2 task one",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "4238e2e9-4a5b-4954-b9b9-6b5909401533",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 2 task two",
          "text": {
            "content": "Day 2 task two",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "ec5b7337-30df-4a7d-a6d3-e313cb097b0c",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "üóìÔ∏è¬†1: Day 2 event",
          "text": {
            "content": "üóìÔ∏è¬†1: Day 2 event",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "aa5f3dec-e418-459b-b5ff-bad3f204977c",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 2 task three",
          "text": {
            "content": "Day 2 task three",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "2a2c118a-9929-4e9c-b054-67ba777e44a0",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "bulleted_list_item",
  },
  {
    "content": {},
    "has_children": false,
    "id": "fb8eb4c1-dc6e-4eb5-bb43-4f6d398d2685",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "divider",
  },
  {
    "content": {
      "title": "Done",
    },
    "has_children": true,
    "id": "941bf143-a1ce-4baf-8d07-fb5328cedc63",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "child_page",
  },
  {
    "content": {},
    "has_children": false,
    "id": "4dae0678-dfb3-4baf-ad2b-ac4865c2169c",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "divider",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": true,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Upcoming",
          "text": {
            "content": "Upcoming",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": true,
    "id": "upcoming-id",
    "parentId": "646656c9-5541-4f87-97de-a186671115fe",
    "type": "toggle",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Something extraneous",
          "text": {
            "content": "Something extraneous",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "dfbd2f12-eeff-4c16-8f72-5570b98a0751",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Fri, 12/24    ‚òîÔ∏è ",
          "text": {
            "content": "Fri, 12/24    ‚òîÔ∏è ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "(84%) ",
          "text": {
            "content": "(84%) ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "blue",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "53¬∞",
          "text": {
            "content": "53¬∞",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": " / ",
          "text": {
            "content": " / ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "orange",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "67¬∞",
          "text": {
            "content": "67¬∞",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "dfbd2f12-eeff-4c16-8f72-5570b98a0751",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 3 task one",
          "text": {
            "content": "Day 3 task one",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "31357e81-a95f-4ef3-996e-40cc1fa7d90b",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 3 task two",
          "text": {
            "content": "Day 3 task two",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "01308fcf-c6c4-41c0-a41d-54e8fbb60101",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 3 task three",
          "text": {
            "content": "Day 3 task three",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "c47fb2b8-4a56-4317-8b55-4824ad449e0a",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Sat, 12/25    ‚òÄÔ∏è ",
          "text": {
            "content": "Sat, 12/25    ‚òÄÔ∏è ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "blue",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "52¬∞",
          "text": {
            "content": "52¬∞",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": " / ",
          "text": {
            "content": " / ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "orange",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "70¬∞",
          "text": {
            "content": "70¬∞",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "7e8b86d0-b8a4-4904-9cbe-a565be11f048",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Day 4 task",
          "text": {
            "content": "Day 4 task",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "last-upcoming-id",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {},
    "has_children": false,
    "id": "2c715df6-e475-4151-8e19-41afab14a367",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "divider",
  },
  {
    "content": {
      "color": "blue",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": "http://localhost:3333/notion/action/key?action=addUpcomingWeek&notionToken=notion-token&upcomingId=upcoming-id",
          "plain_text": "Add following week",
          "text": {
            "content": "Add following week",
            "link": {
              "url": "http://localhost:3333/notion/action/key?action=addUpcomingWeek&notionToken=notion-token&upcomingId=upcoming-id",
            },
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "80d3f02a-8869-4a7a-b406-290641fd1bf2",
    "parentId": "8c8a6ce7-ffe4-46f4-beab-31f1f524b8be",
    "type": "paragraph",
  },
  {
    "content": {},
    "has_children": false,
    "id": "2c715df6-e475-4151-8e19-41afab14a367",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "divider",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "No date yet",
          "text": {
            "content": "No date yet",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "no-date-yet",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "title": "Some child page",
    },
    "has_children": false,
    "id": "no-text",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "child_page",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Mon, 12/27    ‚òÄÔ∏è ",
          "text": {
            "content": "Mon, 12/27    ‚òÄÔ∏è ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "blue",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "52¬∞",
          "text": {
            "content": "52¬∞",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "gray",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": " / ",
          "text": {
            "content": " / ",
            "link": null,
          },
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "orange",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "70¬∞",
          "text": {
            "content": "70¬∞",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "extra-date-1",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "From the future for Monday",
          "text": {
            "content": "From the future for Monday",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "extra-item-1",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "Wed, 12/29",
          "text": {
            "content": "Wed, 12/29",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "extra-date-2",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "paragraph",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "From the future for Wednesday",
          "text": {
            "content": "From the future for Wednesday",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": false,
    "id": "extra-item-2",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {
      "color": "default",
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "href": null,
          "plain_text": "From the future for Wednesday with children",
          "text": {
            "content": "From the future for Wednesday with children",
            "link": null,
          },
          "type": "text",
        },
      ],
    },
    "has_children": true,
    "id": "extra-with-children",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "bulleted_list_item",
  },
  {
    "content": {},
    "has_children": false,
    "id": "c0097bb8-72be-4369-98f8-aaf6fe1ab584",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "divider",
  },
  {
    "content": {
      "title": "Week Template (extra text)",
    },
    "has_children": true,
    "id": "week-template-id",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "child_page",
  },
  {
    "content": {
      "title": "Future",
    },
    "has_children": true,
    "id": "8024ae92-01f4-47f4-8496-0f0db2a95af4",
    "parentId": "b947d72a-b658-4663-816e-3505e4776078",
    "type": "child_page",
  },
]
`;

exports[`lib/notion/quests > POST /notion/quests/:key > adds it at the end if there are somehow no dates and no dividers 1`] = `
"- A new quest
"
`;

exports[`lib/notion/quests > POST /notion/quests/:key > adds it before divider if first date is first item and there is no second date 1`] = `
"- A new quest
"
`;

exports[`lib/notion/quests > POST /notion/quests/:key > adds it before divider if there are no dates 1`] = `
"- A new quest
"
`;

exports[`lib/notion/quests > POST /notion/quests/:key > adds it before second date if first date is first item 1`] = `
"- A new quest
"
`;

exports[`lib/notion/quests > POST /notion/quests/:key > adds quests before next date 1`] = `
"- A new quest
"
`;
